### PR TITLE
[FIX] l10n_it_edi_proxy: Add fallback for remove_signature

### PR DESCRIPTION
**Pull request by lordkrandel**
Communty PR: the import from file on the client also needs the fallback.
Related `iap-apps` PR: odoo/iap-apps#694

Some of the invoices (i.e. those from Servizio Elettrico Nazionale, the ex-monopoly-of-energy company) have custom signatures that rely on an old OpenSSL implementation that breaks the current one that sees them as malformed, so we cannot read those files. Also, we couldn't find an alternative way to use OpenSSL to just get the same result without getting the error.

A new fallback method has been added that reads the ASN1 file structure and takes the encoded pkcs7-data tag content out of it, regardless of the signature.

Being a non-optimized pure Python implementation, it takes about 2x the time than the regular method, so it's better used as a fallback. We didn't use an existing library not to further pollute the dependencies space.

Task link: https://www.odoo.com/web#model=project.task&id=3502910
task-3502910

### CODE REVIEW

File: addons/l10n_it_edi/tools/remove_signature.py [hunk 1]
```diff
@@ -1,11 +1,59 @@
...
+def remove_signature(content, target=None):
+    """ Takes a bytestring supposedly PKCS7 signed and returns its PKCS7-data only """
+    decoded_content = None
+    for removal_strategy in (remove_signature_openssl, remove_signature_fallback):
+        if target:
+            target.remove_signature_method = removal_strategy.__name__
+        try:
+            decoded_content = removal_strategy(content)
+            break
+        except:  # noqa: E722
```
[the remark this answers is not in the document]
Comment (lordkrandel) [reply addons/l10n_it_edi/tools/remove_signature.py:28]: I don't know what OpenSSL may raise, I don't have any recovery option, as it runs on a cron (sometimes). The only thing I can do is log and see if I can use the content anyway.
What would you do instead?

File: addons/l10n_it_edi/tools/remove_signature.py [hunk 2]
```diff
@@ -1,11 +1,59 @@
...
+
+def remove_signature(content, target=None):
+    """ Takes a bytestring supposedly PKCS7 signed and returns its PKCS7-data only """
+    decoded_content = None
+    for removal_strategy in (remove_signature_openssl, remove_signature_fallback):
+        if target:
+            target.remove_signature_method = removal_strategy.__name__
+        try:
+            decoded_content = removal_strategy(content)
+            break
```
[the remark this answers is not in the document]
Comment (lordkrandel) [reply addons/l10n_it_edi/tools/remove_signature.py:27]: I don't remember, I see the arg `content` being updated, I have to check the calls, if everything is fine, I'll remove 👍🏻

**Comment by lordkrandel** [comment]
> What is even the point of `remove_signature_openssl` at this point?
PyOpenSSL is faster and it works in 95% of all cases.
In 5% of cases, the signature is non-standard (old non-strictly standard signature software) and we have to fallback. The fallback is also tested with a smaller number of invoices, whenever the OpenSSL method already fails.

> If there's a readable signature and it's invalid we ignore that, strip the signature, and carry on anyway.
That's what we're doing with both "strategies".

> > It's only to be used as a no-requirements fallback for OpenSSL.
> And I don't understand what "no-requirements" would be in the first place, pyopenssl is part of the core Odoo requirements, and we have inconditional imports right in base
No "additional" requirement to OpenSSL. I could have used other Python libraries, but I went straight full Python instead, so that at least we can debug, improve, update, remove more easily.

**Comment by lordkrandel** [comment]
@xmo-odoo another pass?

File: addons/l10n_it_edi/tools/remove_signature.py [hunk 3]
```diff
@@ -36,16 +78,308 @@ def remove_signature(content):
...
+            raise TypeError("content must be bytes or None")
+        self._content = content
+
+class PrimitiveNode(Asn1Node):
+    """ Primitive Asn1 nodes contain pure data """
+    _content = None
+
+    def _read(self):
+        """ Must be reimplemented in subclasses """
+        raise NotImplementedError(type(self))
```
[the remark this answers is not in the document]
Comment (lordkrandel) [reply addons/l10n_it_edi/tools/remove_signature.py]: I had made some local code to handle other Asn1 nodes and understand stuff, and this dates back to it.

But... you know what? We don't even need that bit, so I cut it all out.

**Comment by lordkrandel** [comment]
@jco-odoo for final review.

File: addons/l10n_it_edi/tools/remove_signature.py [hunk 4]
```diff
@@ -1,11 +1,53 @@
...
 _logger = logging.getLogger(__name__)
 
+
+def remove_signature(content, target=None):
+    """ Takes a bytestring supposedly PKCS7 signed and returns its PKCS7-data only """
+    for removal_strategy in (remove_signature_openssl, remove_signature_fallback):
+        if target:
+            target.remove_signature_method = removal_strategy.__name__
+        with suppress(Exception):
+            return removal_strategy(content)
```
Comment (jco-odoo) [inline addons/l10n_it_edi/tools/remove_signature.py:26]:
```suggestion
            return removal_strategy(content)
        return content
```

**Comment by lordkrandel** [comment]
Done.

**Comment by jco-odoo** [comment]
@robodoo r+

**Comment by lordkrandel** [comment]
@jco-odoo There was a random `hr_payroll` test failing, had to rebase.
Please `r+` again.

**Comment by jco-odoo** [comment]
@robodoo r+

**Comment by jco-odoo** [comment]
@lordkrandel This one is not so random

**Comment by lordkrandel** [comment]
> @lordkrandel This one is not so random

Heh, no, I'm checking. Frak.

**Comment by lordkrandel** [comment]
Aaaaand fixed FYI @jco-odoo .

Here's the diff, as there's a blob and we live in strange times *(feel free to verify 🤣 )*

image

These binary bits that changed are the offsets.
Since I had to change the Codice Fiscale into the test company's Partita IVA to be able to import the file, the binary blocks have moved, and the headers of the blocks with the block length had to be adapted.

<details><summary>Header's formula</summary>
<p>

```
Header's formula for the block's length:
1st byte: there are `n` more bytes (in this case 2)
2nd byte: most significant byte (x >> 8)
3rd byte: less significant byte (y)
[...]

Total length of the block: (x >> 8) + y
```

</p>
</details> 

<details><summary>Test passing</summary>
<p>

image

</p>
</details>

**Comment by lordkrandel** [comment]
1. In case the `remove_signature_fallback` didn't find a header or content, it didn't throw, so empty content was returned as being OK. Fixed by making the method raise in case it doesn't find anything.
2. The decoding method was always trying to remove a CADES signature even from plain XML text. Now we just do that if we have an error in XML parsing the first time, then we re-try.

<details><summary>Here is the diff</summary>
<p>

```py
diff --git a/addons/l10n_it_edi/models/ir_attachment.py b/addons/l10n_it_edi/models/ir_attachment.py
index 12d5d6a048b9..14f991490b32 100644
@@ -19,15 +19,18 @@ class IrAttachment(models.Model):
         """ Decodes a  into a list of one dictionary representing an attachment.
             :returns:           A list with a dictionary.
         """
-        if not (decoded_content := remove_signature(content)):
-            return []
+        def parse_xml(parser, name, content):
+            try:
+                return etree.fromstring(content, parser)
+            except (etree.ParseError, ValueError) as e:
+                _logger.info("XML parsing of %s failed: %s", name, e)
 
         parser = etree.XMLParser(recover=True, resolve_entities=False)
-        try:
-            xml_tree = etree.fromstring(decoded_content, parser)
-        except etree.ParseError as e:
-            _logger.exception("Error when converting the xml content to etree: %s", e)
-            return []
+        if (xml_tree := parse_xml(parser, name, content)) is None:
+            # The file may have a Cades signature, trying to remove it
+            if (xml_tree := parse_xml(parser, name, remove_signature(content))) is None:
+                _logger.info("Italian EDI invoice file %s cannot be decoded.", name)
+                return []
 
         return [{
             'filename': name,
diff --git a/addons/l10n_it_edi/tests/test_edi_import.py b/addons/l10n_it_edi/tests/test_edi_import.py
index 4b5e6a76ec8c..291027a937d9 100644
--- a/addons/l10n_it_edi/tests/test_edi_import.py
+++ b/addons/l10n_it_edi/tests/test_edi_import.py
@@ -9,6 +9,9 @@ from odoo import fields, sql_db, tools, Command
 from odoo.tests import tagged
 from odoo.addons.l10n_it_edi.tests.common import TestItEdi
 
+import logging
+_logger = logging.getLogger(__name__)
+
 
 @tagged('post_install_l10n', 'post_install', '-at_install')
 class TestItEdiImport(TestItEdi):
@@ -42,6 +45,14 @@ class TestItEdiImport(TestItEdi):
     # Vendor bills
     # -----------------------------
 
+    def test_receive_invalid_xml(self):
+        xml_decode = self.env['ir.attachment']._decode_edi_l10n_it_edi
+        with tools.mute_logger("odoo.addons.l10n_it_edi.models.ir_attachment"):
+            self.assertEqual([], xml_decode("none.xml", None))
+            self.assertEqual([], xml_decode("empty.xml", ""))
+            self.assertEqual([], xml_decode("invalid.xml", "invalid"))
+            self.assertEqual([], xml_decode("invalid2.xml", "<invalid/>"))
+
     def test_receive_vendor_bill(self):
         """ Test a sample e-invoice file from
         https://www.fatturapa.gov.it/export/documenti/fatturapa/v1.2/IT01234567890_FPR01.xml
diff --git a/addons/l10n_it_edi/tools/remove_signature.py b/addons/l10n_it_edi/tools/remove_signature.py
index 2790c0db5afd..adfe98a1f82f 100644
--- a/addons/l10n_it_edi/tools/remove_signature.py
+++ b/addons/l10n_it_edi/tools/remove_signature.py
@@ -24,7 +24,6 @@ def remove_signature(content, target=None):
             target.remove_signature_method = removal_strategy.__name__
         with suppress(Exception):
             return removal_strategy(content)
-    return content
 
 # --------------------------------------------------------------------------------
 # UTILS
@@ -117,6 +116,10 @@ def remove_signature_fallback(content):
         elif data_found:
             break
 
+    if not header_found:
+        raise Exception("ASN1 Header not found")
+    if not data_found:
+        raise Exception("ASN1 Content not found")
     return result
 
 # --------------------------------------------------------------------------------
```

</p>
</details>

**Comment by jco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**